RecoEgamma/EgammaTools/interface/MultiToken.h

```diff
@@ -0,0 +1,127 @@
+#ifndef RecoEgamma_EgammaTools_MultiToken_H
+#define RecoEgamma_EgammaTools_MultiToken_H
+
+#include "FWCore/Framework/interface/ConsumesCollector.h"
+#include "FWCore/Utilities/interface/EDGetToken.h"
+
+/*
+ * This class is a wrapper around a vector of EDM tokens, of which at least one
+ * is expected to yield a valid handle.
+ *
+ * The first time you call getValidHandle() or getHandle(), it will go over all
+ * tokens and try to get a valid handle. If no token yields a valid handle, it
+ * will either throw and exception or return the last non-valid handle.
+ *
+ * Once it found a valid handle, it will remember which token was used to get
+ * it and therefore doesn't need to loop over all tokens from there on.
+ *
+ * Example use case: auto-detection of AOD vs. MiniAOD.
+ *
+ * Created by Jonas Rembser on August 3, 2018.
+ */
+
+class MultiToken {
```

Comment:
@Dr15Jones 
would you consider this code to be placed somewhere in FWCore (FWCore/Utilities ?)

The functionality is to pick the first valid product of the same type (including also edm::View) from a fixed list and assume from the first call that this is the product to be used in all following calls.
This is rather common is applications that e.g. need to read the main candidate collection from AOD or from miniAOD, which have a different name and full type of that main candidate collection (PFCandidates vs PackedCandidates)

Reply:
I'm pretty certain this is not what we want people to do. It will force the framework to prefetch everything that was in the list, whether it was used or not. It is by far better, especially for production code, to get the configuration to be what is needed.

We are considering having something equivalent to an EDAlias to allow dynamic determination of an EDProduct based on a specific ordering of lookups. This would be done in such a way as to only prefetch in the given order until sucessful.